File: c++/L2_protected.cpp
#include <iostream>
#include <string>

class Person {
protected:
    std::string name;
};

class Teacher : public Person {
public:
    std::string department;

    void setName(std::string n) {
        name = n; // 'name' is accessible because Teacher inherits from Person
    }

    void print() {
        std::cout << "Name: " << name << ", Department: " << department << std::endl;
    }
};

int main() {
    Teacher teacher1;
    teacher1.setName("Dr. Smith");
    teacher1.department = "Computer Science";
    teacher1.print();

    // The following line would cause a compile error because 'name' is protected
    // std::cout << teacher1.name; // Error
    return 0;
}

File: python/L2_protected.py
class Person:
    def __init__(self, name):
        self._name = name # Protected convention

class Teacher(Person):
    def __init__(self, name, department):
        super().__init__(name)
        self.department = department

    def print_details(self):
        print(f"Name: {self._name}, Department: {self.department}")

teacher1 = Teacher("Dr. Smith", "Computer Science")
teacher1.print_details()

# It's possible to access _name, but the underscore indicates it's not recommended
print(teacher1._name)
